### PR TITLE
adding support for Bytes attribute type 

### PR DESCRIPTION
**Pull request by codeboten**
**Description:** Adding support for Bytes attribute type.

**Link to tracking Issue:** Part of #3535 

This is built on top of https://github.com/open-telemetry/opentelemetry-collector/pull/3740.

### CODE REVIEW

**Comment by tigrannajaryan** [comment]
LGTM.
FYI, @djaglowski (we can use this for raw byte logs once it is merged).

**Comment by bogdandrutu** [comment]
Needs a rebase

**Comment by codeboten** [comment]
@bogdandrutu rebased, looks like some failure in contrib is caused by an update of the go SDK?

File: model/pdata/common.go [hunk 1]
```diff
@@ -116,6 +120,11 @@ func NewAttributeValueArray() AttributeValue {
 	return AttributeValue{orig: &otlpcommon.AnyValue{Value: &otlpcommon.AnyValue_ArrayValue{ArrayValue: &otlpcommon.ArrayValue{}}}}
 }
 
+// NewAttributeValueBytes creates a new AttributeValue with the given []byte value.
+func NewAttributeValueBytes(v []byte) AttributeValue {
+	return AttributeValue{orig: &otlpcommon.AnyValue{Value: &otlpcommon.AnyValue_BytesValue{BytesValue: v}}}
+}
```
Comment (bogdandrutu) [inline model/pdata/common.go:126]: Add a warning about not copying the data, and the data must not be changed after this method is called, or shared between different attributes.

Comment (bogdandrutu) [reply model/pdata/common.go:126]: We try to not allow users to set/get mutable data structures unless controlled by us.

Comment (tigrannajaryan) [reply model/pdata/common.go:126]: I agree with a comment. Note that this is not unique in any way, it is just another reference data type, like the array or the map that we already allow as attribute value.

**Review by tigrannajaryan — APPROVED** [review]
LGTM.

File: model/pdata/common.go (hunk 1 above)
Comment (codeboten) [reply model/pdata/common.go:126]: Right I guess the difference there is with map/array, the NewAttributeValue methods instantiate an empty array/map instead. I'll add a warning.

Comment (codeboten) [reply model/pdata/common.go]: Comment added.

File: model/pdata/common.go [hunk 2]
```diff
@@ -116,6 +120,13 @@ func NewAttributeValueArray() AttributeValue {
 	return AttributeValue{orig: &otlpcommon.AnyValue{Value: &otlpcommon.AnyValue_ArrayValue{ArrayValue: &otlpcommon.ArrayValue{}}}}
 }
 
+// NewAttributeValueBytes creates a new AttributeValue with the given []byte value.
+// The caller must ensure the []byte passed in is not modified after the call is made, sharing the data
+// across multiple attributes is discouraged.
```
Comment (bogdandrutu) [inline model/pdata/common.go:125]: is forbidden :) probably. We need to have a stronger comment

File: model/pdata/common.go [hunk 3]
```diff
@@ -221,6 +241,13 @@ func (a AttributeValue) SetBoolVal(v bool) {
 	a.orig.Value = &otlpcommon.AnyValue_BoolValue{BoolValue: v}
 }
 
+// SetBytesVal replaces the []byte value associated with this AttributeValue,
+// it also changes the type to be AttributeValueTypeBytes.
+// Calling this function on zero-initialized AttributeValue will cause a panic.
+func (a AttributeValue) SetBytesVal(v []byte) {
```
Comment (bogdandrutu) [inline model/pdata/common.go:247]: Same comment here.

File: model/pdata/common.go [hunk 4]
```diff
@@ -193,6 +206,13 @@ func (a AttributeValue) ArrayVal() AnyValueArray {
 	return newAnyValueArray(&arr.Values)
 }
 
+// BytesVal returns the []byte value associated with this AttributeValue.
+// If the Type() is not AttributeValueTypeBytes then returns false.
+// Calling this function on zero-initialized AttributeValue will cause a panic.
+func (a AttributeValue) BytesVal() []byte {
```
Comment (bogdandrutu) [inline model/pdata/common.go:212]: Comment that it is discouraged to modify the array in-place.

Comment (codeboten) [reply model/pdata/common.go:213]: Added.

File: model/pdata/common.go (hunk 3 above)
Comment (codeboten) [reply model/pdata/common.go:250]: 👍

File: model/pdata/common.go (hunk 2 above)
Comment (codeboten) [reply model/pdata/common.go]: 👍

File: model/pdata/common.go [hunk 5]
```diff
@@ -193,6 +206,14 @@ func (a AttributeValue) ArrayVal() AnyValueArray {
 	return newAnyValueArray(&arr.Values)
 }
 
+// BytesVal returns the []byte value associated with this AttributeValue.
+// If the Type() is not AttributeValueTypeBytes then returns false.
+// Calling this function on zero-initialized AttributeValue will cause a panic.
+// Modifying the returned []byte in-place is strongly discouraged.
```
Comment (bogdandrutu) [inline model/pdata/common.go:212]: let's use "forbidden" here as well, be more strict initially then we can relax if needed.

File: model/pdata/common.go [hunk 6]
```diff
@@ -503,6 +542,14 @@ func (am AttributeMap) InsertBool(k string, v bool) {
 	}
 }
 
+// InsertBytes adds the []byte Value to the map when the key does not exist.
+// No action is applied to the map where the key already exists.
+func (am AttributeMap) InsertBytes(k string, v []byte) {
```
Comment (bogdandrutu) [inline model/pdata/common.go:547]: Same comment should be on all these as we have for Set/New.

**Review by bogdandrutu — APPROVED** [review]

**Review by bogdandrutu — APPROVED** [review]

**Pull request merged**